// gas-sheetdb/lib/RowsList.js
/**
 * @fileoverview RowsList class : sub module of SheetDb
 * @module lib/RowsList
 */

'use strict';

var assert = require('./assert'),
    SheetDb = require('./SheetDb');


/**
 * RowsList class
 * @constructor
 * @extends {SheetDb}
 * 
 * @param {SheetDb} sheetdb
 */
function RowsList(sheetdb, func) {

  if (! sheetdb) {
    throw new Error('RowsList create: missing sheetdb.');
  }

  this.parent_ = sheetdb;
  this.rowsList_ = [];

  if (func) {
    var max = this.parent_.getNumRows();
    for (var row = 1; row <= max; row++) {
      if (func(this.parent_, row)) {
        this.rowsList_.push(row);
      }
    }
  }
  return this;
}

RowsList.prototype = new SheetDb(RowsList);

RowsList.prototype.getNumRows = function() {
  return this.rowsList_.length;
};

RowsList.prototype.getRow = function(num) {
  return this.rowsList_[num - 1];
};

RowsList.prototype.getRows = function() {
  return this.rowsList_.concat();
};

RowsList.prototype.select = function(f) {
  return new RowsList(this, f);
};

module.exports = RowsList;
